### PR TITLE
Security: Update default CSP template and fix firefox CSP issues

### LINKED ISSUE #33323: Frontend: Webpack injected scripts don't include CSP nonce
**What happened**:
When enabling [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP), Safari won't load Grafana since some dynamically loaded JS bundles don't have the "nonce" attribute. The JS loading errors can look as follows:

```
[Error] Refused to load http://localhost:3000/public/build/default~DashboardPage~SoloPanelPage.7bb0ed519991a573201e.js because it does not appear in the script-src directive of the Content Security Policy.
```

AFAICT this is happening because Webpack injects scripts without the nonce attribute, which is necessary for JavaScript to be loadable when CSP is enabled (provided that `$NONCE` is in the `script-src` setting).

**What you expected to happen**:
For all JS bundles to have the nonce attribute set.

**How to reproduce it (as minimally and precisely as possible)**:
1. Enable CSP
2. Launch Grafana
3. Log into Grafana in Safari

This should break visibly.

**Anything else we need to know?**:

*When this is fixed, `script-src` changes in #34363 should be reverted, since it's a workaround that should presumably no longer be required*.

**Environment**:
- Grafana version:
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser: OSX / Safari
- Grafana plugins:
- Others:

### LINKED ISSUE #34897: Grafana 8.0-beta2 fails to load in Safari.
**What happened**:

Logged in on Safari Version 14.0.1 (16610.2.8.1.1) on my Mac, and the side menu is visible for a split second before it crashes with the "If you're seeing this Grafana has failed to load its application files" fallback error message.

Screen Shot 2021-05-28 at 2 01 06 pm

 
**What you expected to happen**:

Grafana loads!

**How to reproduce it (as minimally and precisely as possible)**:

Open https://g80-beta.grafana.net/?orgId=1 on Safari (Version 14.0.1 (16610.2.8.1.1)).

**Anything else we need to know?**:

Hosted Grafana works fine.

It looks like the relevant exception in the console is:

```
TypeError: undefined is not an object (evaluating 'this.window.panelsRendered=this.panelsRendered')
    renderingCompleted — Counter.tsx:8
    renderPanel — DashboardPanel.tsx:140
```

There's also bunch of other errors, which may or may not be related, around security and CSP?:

 - Refused to connect to wss://g80-beta.grafana.net/api/live/ws because it does not appear in the connect-src directive of the Content Security Policy.
 - The operation is insecure. https://g80-beta.grafana.net/public/build/vendors~app.4f857895d25408eb0dd2.js:2
 - Refused to load https://fonts.gstatic.com/s/robotomono/...woff2 because it does not appear in the font-src directive of the Content Security Policy.

@ryantxu @wbrowne Maybe you're interested in the web socket and CSP stuff?

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Fixes various CSP related issues:
- Add 'strict-dynamic' back to script-src
- Add ws(s)://$DOMAIN:$PORT to connect-src
- Change onEvent to on-event in angular templates to fix CSP issues in firefox.
- Add blob: to style-src

**Which issue(s) this PR fixes**:
Fixes #33323
Fixes #34897

### CODE REVIEW

**Comment by torkelo** [comment]
Awesome! think we should not add fonts.gstatic.com  , just download the files and place them in https://github.com/grafana/grafana/tree/main/public/fonts/roboto

**Review by torkelo — COMMENTED** [review]
for ws://$DOMAIN:$PORT wss://$DOMAIN:$PORT; do we need to use domain & port? what is you use Grafana behind a proxy? think it's better to 
use ws://$ROOT_WITH_PROTOCOL  and replace that with root_url (without http or https)

**Comment by kaydelaney** [comment]
> for ws://$DOMAIN:$PORT wss://$DOMAIN:$PORT; do we need to use domain & port? what is you use Grafana behind a proxy? think it's better to
> use ws://$ROOT_WITH_PROTOCOL and replace that with root_url (without http or https)

Are the values for `$(domain)` and `$(http_port)` as used in the value for `root_url` (root_url = %(protocol)s://%(domain)s:%(http_port)s/) not the same as for `cfg.Domain` and `cfg.HTTPPort`?

**Comment by torkelo** [comment]
@kaydelaney no not necessarily that is just the default template, http port is what grafana server listens to, does not need to be the same as the public facing url / port (root_url)

**Comment by kaydelaney** [comment]
> @kaydelaney no not necessarily that is just the default template, http port is what grafana server listens to, does not need to be the same as the public facing url / port (root_url)

Ah, right, of course. Made a change, let me know if it looks okay.

**Review by torkelo — APPROVED** [review]
Looks great, awesome job on this! see you found a solution to the onChange issue! 

only nit is $ROOT_URL naming in the template since it is not an URL really (missing protocol), $PUBLIC_DOMAIN_AND_PATH ?

**Review by jackw — APPROVED** [review]
Amazing work on this @kaydelaney. 👏👏👏

**Review by dprokop — APPROVED** [review]
Impressive work @kaydelaney!

File: pkg/middleware/csp.go [hunk 1]
```diff
@@ -42,6 +43,10 @@ func AddCSPHeader(cfg *setting.Cfg, logger log.Logger) macaron.Handler {
 
 		nonce := base64.RawStdEncoding.EncodeToString(buf[:])
 		val := strings.ReplaceAll(cfg.CSPTemplate, "$NONCE", fmt.Sprintf("'nonce-%s'", nonce))
+
+		re := regexp.MustCompile(`^\w+:(//)?`)
+		rootURL := re.ReplaceAllString(cfg.AppURL, "")
+		val = strings.ReplaceAll(val, "$ROOT_URL", rootURL)
```
Comment (aknuds1) [inline pkg/middleware/csp.go:49]: Shouldn't this be named `rootPath`/`ROOT_PATH`, since it's no longer a URL when you strip the protocol?

**Review by aknuds1 — CHANGES_REQUESTED** [review]
I think the name ROOT_URL is misleading, since it isn't a URL

**Review by aknuds1 — APPROVED** [review]

File: pkg/middleware/csp.go [hunk 2]
```diff
@@ -42,6 +43,10 @@ func AddCSPHeader(cfg *setting.Cfg, logger log.Logger) macaron.Handler {
 
 		nonce := base64.RawStdEncoding.EncodeToString(buf[:])
 		val := strings.ReplaceAll(cfg.CSPTemplate, "$NONCE", fmt.Sprintf("'nonce-%s'", nonce))
+
+		re := regexp.MustCompile(`^\w+:(//)?`)
+		rootURL := re.ReplaceAllString(cfg.AppURL, "")
+		val = strings.ReplaceAll(val, "$ROOT_PATH", rootURL)
```
Comment (aknuds1) [inline pkg/middleware/csp.go:49]:
```suggestion
		rootPath := re.ReplaceAllString(cfg.AppURL, "")
		val = strings.ReplaceAll(val, "$ROOT_PATH", rootPath)
```

**Review by aknuds1 — COMMENTED** [review]
`rootURL` should still be renamed to `rootPath`.

**Comment by hugohaggmark** [comment]
Amazing work on this @kaydelaney 🎉

**Comment by hugohaggmark** [comment]
@kaydelaney friendly reminder to add backport, approve the backport and merge it. Thank you!

**Pull request merged**